### ISSUE TITLE
Get time and date from server

### ISSUE THREAD

**Issue by smaragdus** [filecxx/FileCentipede#13]
It seems that [File Centipede](http://filecxx.com/en_US/index.html) lacks a common feature available in most download managers

- **Get date and time of file from server**

which means that the original timestamps of the downloaded files are not preserved.

Several examples of this feature in other download managers:

- [Ant Download Manager](https://antdownloadmanager.com/) (commercial):

`Sets the web server datetime for the downloaded file`

**screen:**

![Ant Download Manager 2 2 4 Build 77918 - 2021-10-07 - date-time - 002](https://user-images.githubusercontent.com/9025757/157998504-7d114536-503f-45b9-99ec-8a4b579227b0.png)

- [DownloadStudio](http://www.conceiva.com/products/downloadstudio/default.asp) (commercial):

`Use web server's date and time for downloaded files`

![DownloadStudio 10 0 4 0 - 2021-10-07 - date-time - 002](https://user-images.githubusercontent.com/9025757/157998644-3bd3e2dc-ea55-4d5f-af04-a6701fe7d960.png)

- [DownZemAll!](https://setvisible.github.io/DownZemAll/) (free, open source)

```
Get time from server for the file's...:

Last modified time

Creation time (may not be not supported on UNIX)

Most recent access (e.g. read or written to)

Metadata change time
```
**screen:**

![DownZemAll! 2 4 3 - 2022-03-12 - time-date - 002](https://user-images.githubusercontent.com/9025757/157998918-e3d322e6-e8ec-4899-8eee-e9afc90dcfcc.png)

- [FireDM](https://github.com/firedm/FireDM) (free, open source):

`Use server timestamp for downloaded files`

**screen:**

![FireDM 2021 12 26 - 2022-03-12 - date-time - 002](https://user-images.githubusercontent.com/9025757/157999014-bb6644b9-ac1b-45aa-8ef8-d959953e1173.png)

- [Free Download Manager](https://www.freedownloadmanager.org/) (free, closed source):

`Set the date of a file identical to the date specified by the server`

**screen:**

![Free Download Manager 3 9 7 Build 1627 - 2021-10-07 - date-time - 002](https://user-images.githubusercontent.com/9025757/157998800-e5372183-0900-41d0-b457-14b19ff23b39.png)

- [HTTP Downloader](https://erickutcher.github.io/#HTTP_Downloader) (free, open source):

`Set date and time of file from server response`

**screen:**

![HTTP Downloader 1 0 4 5 - 2021-10-07 - date-time - 002](https://user-images.githubusercontent.com/9025757/157999166-49f9f3da-4f91-46ae-888e-f6e12ad45af4.png)

- [Internet Download Accelerator](https://westbyte.com/ida/) (commercial):

`Get date and time of file creation from the server`

**screen:**

![Internet Download Accelerator 6 13 1 1557 - 2021-10-07 - time-date - 002](https://user-images.githubusercontent.com/9025757/157999255-a1a73767-0233-4ca8-a454-d8e691c48bf4.png)

The command-line download manager [aria2](https://aria2.github.io/) also supports [this feature](https://aria2.github.io/manual/en/html/aria2c.html#cmdoption-r) via the following parameter:

`--remote-time`

So download managers which use [aria2](https://github.com/aria2/aria2) as download engine also support this feature.

A couple of examples:

- [AWGG](https://sourceforge.net/projects/awgg/) (free, open source):

`--remote-time`

**screen:**

![AWGG 0 6 0 5953 - 2021-10-08 - date-time - 002](https://user-images.githubusercontent.com/9025757/157999396-66494427-2a6a-459a-9476-f45134a53032.png)

- [uGet](https://sourceforge.net/projects/urlget/) (free, open source):

`--remote-time`

**screen:**

![uGet 2 2 3-2 - 2021-10-08 - date-time - 002](https://user-images.githubusercontent.com/9025757/157999461-b54bc2fd-f2e9-4abf-aa73-47c74a094e92.png)

I suggest that a new check-box be added in program settings:

- **Get date and time of file from server**

or something like that so the files downloaded with [File Centipede](https://github.com/filecxx/FileCentipede) will preserve the original timestamps received from server.

I hope that this feature will be implemented as I prefer the downloaded files to retain their original timestamps.

Regards

**Comment by filecxx** [comment]
http engine current is getting etag and last-modified fields from HTTP response header, if no etag, 

![image](https://user-images.githubusercontent.com/100348948/158001313-ff4b18ca-6d3f-4ea0-9dc9-efd758e7593d.png)

the last-modified field will be used.

![image](https://user-images.githubusercontent.com/100348948/158001334-420cfe15-0262-4aec-9ac2-d4f239dae17e.png)

this is used to ignore that, which means the server changed file identifier, download anyway.

if it's unchecked, the task state display:

![image](https://user-images.githubusercontent.com/100348948/158001433-2bba25a6-f773-412c-b5a6-34f524a70825.png)

I'm not sure if we're talking about the same stuff, or you meant  
![image](https://user-images.githubusercontent.com/100348948/158001557-ee06e5dc-0a7f-48bc-8f8d-c79ee0cdabfa.png)

**Comment by smaragdus** [comment]
@filecxx 

I will try to explain what I mean using a couple of examples.

- First- [EMDB](https://www.emdb.eu/downloads.html) download ([direct download link](https://www.emdb.eu/bin/emdb.zip)):

1. Timestamp - FireDM - 23-02-2022 11:09:22 - original timestamp from server
2. Timestamp - HTTP Downloader - 23-02-2022 11:09:22 - original timestamp from server
3. Timestamp - File Centipede - 13-03-2022 01:07:25 - timestamp equal to when file was downloaded

**screen:**

![EMDB - 2022-03-13 - timestamp - 006](https://user-images.githubusercontent.com/9025757/158039018-5280cd2d-9129-4163-974e-df84b70c472e.png)

- Second- [File Centipede](https://github.com/filecxx/FileCentipede/releases) download ([direct download link](https://github.com/filecxx/FileCentipede/releases/download/1.xx/filecxx_1.0_win_x64.zip)):

1. Timestamp - FireDM - 12-03-2022 05:20:48 - original timestamp from server
2. Timestamp - HTTP Downloader - 12-03-2022 05:20:48 - original timestamp from server
3. Timestamp - File Centipede - 13-03-2022 01:35:42 - timestamp equal to when file was downloaded

**screen:**

![File Centipede - 2022-03-13 - timestamp - 003](https://user-images.githubusercontent.com/9025757/158039096-545bcd73-127e-4db4-8713-d27442894d66.png)

As seen in the examples above [FireDM](https://github.com/firedm/FireDM) and [HTTP Downloader](https://github.com/erickutcher/httpdownloader) can retain the original timestamp of the downloaded file- the timestamp of the downloaded file is equal to the timestamp received from server response.

While the files downloaded with [File Centipede](https://github.com/filecxx/FileCentipede) do not retain the original timestamp received from server response- the timestamps of the downloaded files are equal to the time when the files were downloaded.

Enabling or disabling **Ignore remote modified** option did not make any difference in my tests. However I see that there is a new release with the same version number which I have not yet tested.

In short- I would like that there is a persistent option in **File Centipede** settings which allows all the downloaded files to keep the original timestamps as received from server response.

In short- if file **xyz.zip** is uploaded on **2022-03-13 02:09:14** I want the downloaded file with **File Centipede** to have the same timestamp- **2022-03-13 02:09:14**.

I hope I was able to make it clear what I mean.

**Comment by filecxx** [comment]
I got you, but not all site's HTTP header contains  "creation time" and "last modified time", so the "new checkbox" name should be:  "use the timestamp from webserver as file creation time (if exists) ??", I prefer a shorter name.

**Comment by smaragdus** [comment]
@filecxx 

> I got you, but not all site's HTTP header contains "creation time" and "last modified time", so the "new checkbox" name should be: "use the timestamp from webserver as file creation time (if exists) ??", I prefer a shorter name.

I think that most servers have this header,

About the check box- I too prefer shorter names for options. Perhaps the check box might be:

- **Get file timestamp from server**

It is brief and understandable enough. In the screens above there were longer and more explanatory lines but I suppose that even the short line as the one above will be clear enough what it is all about.

**Comment by smaragdus** [comment]
@filecxx 

> Sorry mate, "use the timestamp from webserver as file creation time (if exists) ??" haven't been added into version 1.1, but I will do it in the future, I promise.

I am not in a hurry at all. I got the latest version (1.1) and when I have time (I hope soon) I will do more tests and I will most likely open new issues (feature requests).

Actually I have more than enough download managers (both free and commercial) but I see great potential in File Centipede (I was really impressed when I first tested the initial version) so I will continue to test it and suggest possible enhancements and improvements.

I really appreciate your work and dedication to this project!

Regards

**Comment by filecxx** [comment]
Try 1.5, if it's not good enough, maybe I can refine the stuff.

**Comment by smaragdus** [comment]
@filecxx 

Good job, thanks!

I will suggest a tiny enhancement. Below I will try to explain what I mean.

For testing I used [File Centipede](http://filecxx.com/en_US/index.html) [version](https://sourceforge.net/projects/filecentipede/files/) [1.5](https://sourceforge.net/projects/filecentipede/files/1.5/) from [SourceForge](https://sourceforge.net/projects/filecentipede/) (**filecxx_1.5_win_x64.zip**).

- When using [HTTP Downloader](https://erickutcher.github.io/#HTTP_Downloader) and [FireDM](https://github.com/firedm/FireDM) **Last Modified Date** is equal to **Creation Date**.
- When using [File Centipede](http://filecxx.com/en_US/index.html) the **Creation Date** is equal to when **file was downloaded**.
- It would be better if **Last Modified Date** and **Creation Date** match when 'Last modified time' is set to 'Remote' in Settings.

**screen:**
![File Centipede 1 5 - 2022-03-23 - timestamp - 003](https://user-images.githubusercontent.com/9025757/159594674-a7209682-d660-41a7-b294-69aed060e057.png)

Is this little fine-tuning is done the implementation of **Last modified time** option would be perfect!

Thank you for your excellent work on this project!

Regards

**Comment by filecxx** [comment]
Since "creation date" is windows only, I would like to add it into the next major version, 2.0, and in the next major version a new tab "network files" will be added, it's a file manager for FTP, SSH, WebDAV, and HTTP, I wish it could be better than xshell and winscp, if you have any recommendation, please let me know!

Thank you for helping with this project.

Regards

**Comment by smaragdus** [comment]
@filecxx 

Thanks for planning to implement what I suggested.

Site manager is a common feature but most download managers usually have rather basic FTP support, they just support FTP protocol downloads.

The download manager which I know to have FTP explorer module is commercial [Internet Download Accelerator](https://westbyte.com/ida/). screen:

![Internet Download Accelerator 6 13 1 1557 - 2022-03-24 - 001](https://user-images.githubusercontent.com/9025757/159832272-851a867c-1293-43cc-b532-5ba3be39344d.png)

Wouldn't it be too much work to add FTP manager? This would mean that [File Centipede](http://filecxx.com/en_US/index.html) would become a FTP client too. WebDAV support would be a unique feature. Do you mean that the users would be able to upload files too?

I have not yet offered any real help. Perhaps I may suggest some enhancements in the GUI department as I have used lots of download managers and know what features are handy and useful. When I have time I will most likely open some more issues.

When the developers are helpful like you it is pleasure to contribute with ideas.

Keep the excellent work!
